### ISSUE TITLE
Refactor `OCPBalance` to align with `DataFit`

### ISSUE THREAD

**Issue by brosaplanella** [paramm-team/pybamm-param#42]
At the moment `OCPBalance` is very different from `DataFit` which makes it hard to maintain. It should be reformatted so it is more similar (e.g. in the use of the `parameters_optimise`, `cost_function_parameters`, etc.).

**Comment by brosaplanella** [comment]
Just realised that many things will change once #32 is merged, so might be better to do that first. If you want, you can review that PR first. I have now added you to the team so you should be able to leave a review.

**Comment by PipGrylls** [comment]
Ah thanks for the heads up I will have a look

**Comment by PipGrylls** [comment]
Changes pulled from main (post main updated with #32) https://github.com/paramm-team/pybamm-param/commit/e3e20ff5388a8e000baaa0b8669930d1f47e1264

**Comment by PipGrylls** [comment]
In the base class `__init__` we define the following variables:

``` 
def __init__(self):
        
        """
        Initialize the class properties

        x0 : array-like
            Initial guess for the optimization problem
        bounds : tuple
            bounds of the optimization problem
        scalings : array-like
            scalings for the optimization problem
        """
        self.x0 = None
        self.bounds = None
        self.scalings = None
```

OCPBalance does use `super().init()`, data fit does not. 

We then have some disconnect between the names of input data and reference data. 

I will try and pull these things together, to fulfil the brief of this issue but additionally I think there needs to be some strict templating applied to how further optimisation problems are defined.

We can choose the direction the templating goes in but making it permissive will make long term upkeep harder due to edge cases.

**Comment by brosaplanella** [comment]
I have revisited the two classes (plus the new `GITT` one we are starting to develop) and they all have the following common features. The first column are the abstract concepts needed to define the class, while the value in each cell is how these are passed/defined for the optimisation problem of the respective column.

|               	| DataFit                    | OCPBalance   | GITT 	                |
|---------------	|---------	                 |------------	  |------	                |
| model         	| `simulation`        	 |`data_fit`          |choose from given|
| data          	|`data`         	         |`data_ref`         |`data`                     |
| parameters    |`model_parameters` |fixed            	  |fixed       	        |
| variables fit 	|`variables_optimise` |fixed            	  |fixed       	        |
| cost function |`cost_function`         |`cost_function`|`cost_function` 	|

The various classes might have inputs that are specific to them (e.g. GITT will require some parameters).

Then, the `__init__` would have to:
- Process/assemble model
- Process/clean data
- Assemble parameters (including those from cost function)

The specific methods will be different for each class, but we could just call `process_model`, etc. from the base class and define what the function does in each subclass.

**Comment by PipGrylls** [comment]
Thats really useful thanks.

I think there is some argument to include some sphinx docs (new issue needed for this), pybamm uses them so it would be consistent.

**Comment by brosaplanella** [comment]
There are some docs here (https://pybamm-param.readthedocs.io/en/latest/?badge=latest) but there is a lot that can be improved.
